### ISSUE TITLE
External download links are not working as it should

### ISSUE THREAD

**Issue by rynaldos** [woocommerce/woocommerce#26879]
Hi. 

**Prerequisites (mark completed items with an [x]):**
- [ x] I have have carried out troubleshooting steps and I believe I have found a bug.
- [ x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**

Downloadable files that are linked to an external source, such as Google drive does not download correctly post-purchase. 

The file in question is here: https://drive.google.com/uc?export=download&id=1q4y4ZDZAhX1chmJRoPpF3Ag12waPZtG2 and can be downloaded correctly, but as soon as this is added to the file URL here: 

<img width="830" alt="Screenshot 2020-06-26 at 09 38 46" src="https://user-images.githubusercontent.com/6198774/85832728-e433e000-b790-11ea-9613-e4d8ca395fcc.png">

It then downloads a file that look like this: 

<img width="642" alt="Screenshot 2020-06-26 at 09 39 58" src="https://user-images.githubusercontent.com/6198774/85832808-07f72600-b791-11ea-98c7-589fceb446d1.png">

**Expected behavior**

I would expect the same .pdf file to be downloaded correctly. 

**Actual behavior**

The file is download in a weird format as seen above. 

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**

Steps to reproduce the bug:

1. Create a downloadable product and make it free
2. Add a file URL to an external source (can use the link above)
3. Check out with this product
4. Download the file from the `order-received` page. 

**Screenshots**

See above.

**Isolating the problem (mark completed items with an [x]):**

- [x ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x ] I can reproduce this bug consistently using the steps above.

**Comment by tammullen** [comment]
Hi @rynaldos 

Thank you for taking the time raising this issue, we really appreciate your help. 

The reason this happens is because the file you have linked to is served by Google Drive.
For Google Drive the name of the file is not transmitted in the url so the file name returns as “uc”. 

There are some workarounds for this

- If the file is manually renamed after download to have the extension `.pdf` it should be able to be opened as expected.
- If you go to WooCommerce / Settings / Products / Downloadable products and change the File Download Method to `Redirect only (insecure)` the file will download with the original file name ok.  Please see here for more details on what the Download Methods mean https://docs.woocommerce.com/document/digital-downloadable-product-handling/#section-1
![Screenshot 2020-06-30 at 16 31 22](https://user-images.githubusercontent.com/24649833/86146220-d9aa7b00-baef-11ea-91ac-9129c1a75d69.png)
- See this forum post for another possible workaround https://wordpress.org/support/topic/download-problem-10/

I am marking this as a bug as when the Download Method is set to `Redirect only (insecure)` the following notice is displayed.
![Screenshot 2020-06-30 at 16 37 27](https://user-images.githubusercontent.com/24649833/86146354-0494cf00-baf0-11ea-88f5-06bd8b213e19.png)
This notice mentions for Google Drive the right method will be automatically used and we should select a different download method. However as you have noted in your issue when using one of the other methods with a Google Drive URL the file name is incorrect.

​
We won’t be able to include this fix in the upcoming release due to the lower priority of this issue compared to others reported. We’re going to add it to our backlog so we can include it in our planning for one of our future releases.

**Comment by m155y** [comment]
Another report in 10457082-hc

**Comment by sukafia** [comment]
I was just playing around with downloadable products and noticed this issue on my end.

**Comment by posegate** [comment]
I ran into this too, and it is problematic.

If I edit the class-wc-download-handler.php at line 214, I can get it to work.

`		if ( strstr( $filename, '?' ) ) {
			$filename = current( explode( '?', $filename ) );
			
			$product = wc_get_product( $product_id );
			$drive_filename = $product->get_downloads();
			foreach( $drive_filename as $key => $each_download){
				$filename = $each_download["name"];
			}
		}
`

**Comment by biguenique** [comment]
This Google Drive "uc" URL actually redirects to a download URL, which sends the file content along with these headers (among others):

```http
content-disposition: attachment;filename="Understanding Psychology (10th Ed)(gnv64).pdf";filename*=UTF-8''Understanding%20Psychology%20%2810th%20Ed%29%28gnv64%29.pdf
content-type: application/pdf
```

@tammulen WooCommerce should relay the value of `filename` in the `Content-Disposition` header when available as the downloaded file name, instead of relying on the path of the URL, which is only meaningful when files are served directly.

**Comment by Ali-RS** [comment]
Hi!

Any plan to fix this issue?

**Comment by dimitrow7** [comment]
Any plan to fix this?

**Comment by chadamski** [comment]
Hey, team! Is there an update on this issue? In the era of large digital downloads that need to be hosted externally from WordPress, the current behavior seems like a critical limitation...

**Comment by ayeronnie** [comment]
> Hey, team! Is there an update on this issue? In the era of large digital downloads that need to be hosted externally from WordPress, the current behavior seems like a critical limitation...

Exactly. I've been hoping for a fix for this for a long time. I am currently stuck at Woo Version 5.4.3 because it was the last version I found to still work. It's very frustrating because this use to work and now it's broken. Yes, I have gigs of digital files, and AWS is so much more cost efficient. I also can not update other plugins because of this issue, including my payment gateway. This is hurting my biz. I love Woocommerce, and it's not like I can switch carts.

Woo has a plugin for S3, but it does not look like it can convert products. I have way too many products to change by hand. 

Can others "bump" this, to maybe get some attention?? Heck, I am willing to pay for a solution. So need this fixed.

**Comment by saniyasaher20** [comment]
Please fix this issue as soon as possible, because its a serious issue for all eCommerce users.

**Comment by saniyasaher20** [comment]
If anyone here have found any solution then please comment here.

**Comment by anastas10s-afk** [comment]
8309950-zen

**Comment by maximilliangeorge** [comment]
This is a hack solution, but you can add the extension manually like so:

```php
add_filter('woocommerce_file_download_filename', function ($filename) {
    return $filename . '.zip';
});
```

**Comment by pmcpinto** [comment]
It looks like this is still relevant. I've marked it as ready for engineering triage.

Current evidence: The current download handler still derives the download filename from basename(file URL), strips query parameters, and then sends that value in Content-Disposition. For Google Drive style URLs like `uc?export=download&id=...`, trunk still resolves the filename to `uc`, so the reported failure remains plausible on current core when using non-redirect download methods.
